### PR TITLE
Extend caldav resource search

### PR DESCRIPTION
**Pull request by miaulalala**
Ref https://github.com/nextcloud/calendar/issues/3184

### CODE REVIEW

File: apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php [hunk 1]
```diff
@@ -340,35 +362,87 @@ public function searchPrincipals($prefixPath, array $searchProperties, $test = '
...
+	/**
+	 * @param $query
+	 * @param $usersGroups
+	 * @return array
+	 */
+	private function getRows($query, $usersGroups):array {
+		try {
+			$stmt = $query->executeQuery();
+		} catch (Exception $e) {
+			$this->logger->error($e->getMessage());
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:417]:
```suggestion
			$this->logger->error("Could not search resources: " . $e->getMessage(), ['exception' => $e]);
```

File: apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php [hunk 2]
```diff
@@ -432,7 +506,7 @@ public function findByUri($uri, $principalPrefix) {
 	 *
 	 * @param String[] $row
 	 * @param String[] $metadata
-	 * @return Array
+	 * @return array
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:509]: can we haz types? `string[]` or whatever these are?

**Review by ChristophWurst — COMMENTED** [review]
Could you provide us with an example curl with which we can test this search? :)

**Comment by miaulalala** [comment]
Test search cURL request:

```bash
curl -k -v http://YOURINSTANCEURL -X REPORT -u username:password --data-raw '<?xml version="1.0"?><x0:principal-property-search xmlns:x0="DAV:" test="allof"><x0:property-search><x0:prop><x2:room-features xmlns:x2="http://nextcloud.com/ns"/></x0:prop><x0:match>PHONE,TV</x0:match></x0:property-search><x0:property-search><x0:prop><x2:room-seating-capacity xmlns:x2="http://nextcloud.com/ns"/></x0:prop><x0:match>100</x0:match></x0:property-search><x0:prop><x1:calendar-user-type xmlns:x1="urn:ietf:params:xml:ns:caldav"/><x0:principal-URL/><x0:alternate-URI-set/><x2:room-features xmlns:x2="http://nextcloud.com/ns"/><x2:room-seating-capacity xmlns:x2="http://nextcloud.com/ns"/></x0:prop><x0:apply-to-principal-collection-set/></x0:principal-property-search>'  -H 'Content-Type: application/xml; charset=UTF-8' -H 'Depth: 0'
```

This search will  look for a room with `feature TV, PHONE` and `seating capacity >= 100`

Of course you'll only find results if you have rooms that have these properties.

The room properties are stored in `calendar_rooms_md`.

**Review by st3iny — APPROVED** [review]
I tested it via my improved search frontend and it worked well. Code looks good too.

File: apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php [hunk 3]
```diff
@@ -340,35 +362,87 @@ public function searchPrincipals($prefixPath, array $searchProperties, $test = '
 	 *
 	 * @param $key
 	 * @param $value
+	 * @param array $usersGroups
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:365]:
```suggestion
	 * @param string[] $usersGroups
```

Comment (ChristophWurst) [reply apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:365]: $key can also get a `string` typehint I guess

File: apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php [hunk 4]
```diff
@@ -340,35 +362,87 @@ public function searchPrincipals($prefixPath, array $searchProperties, $test = '
...
+	 * Searches principals based on room features
+	 * e.g.:
+	 * '{http://nextcloud.com/ns}room-features' => 'TV,PROJECTOR'
+	 *
+	 * @param $key
+	 * @param $value
+	 * @param array $usersGroups
+	 * @return array
+	 */
+	private function searchPrincipalsByRoomFeature($key, $value, $usersGroups = []): array {
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:384]:
```suggestion
	private function searchPrincipalsByRoomFeature(string $key, string $value, array $usersGroups = []): array {
```

File: apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php [hunk 5]
```diff
@@ -340,35 +362,87 @@ public function searchPrincipals($prefixPath, array $searchProperties, $test = '
...
+	 * Searches principals based on room seating capacity or vehicle capacity
+	 * e.g.:
+	 * '{http://nextcloud.com/ns}room-seating-capacity' => '100'
+	 *
+	 * @param $key
+	 * @param $value
+	 * @param array $usersGroups
+	 * @return array
+	 */
+	private function searchPrincipalsByCapacity($key, $value, $usersGroups = []):array {
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:402]:
```suggestion
	private function searchPrincipalsByCapacity(string $key, string $value, array $usersGroups = []):array {
```

File: apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php [hunk 6]
```diff
@@ -432,7 +506,7 @@ public function findByUri($uri, $principalPrefix) {
 	 *
 	 * @param String[] $row
 	 * @param String[] $metadata
-	 * @return Array
+	 * @return String[]
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:509]:
```suggestion
	 * @return string[]
```

Comment (ChristophWurst) [reply apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:509]: you can also fix the rows above :)

File: apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php [hunk 7]
```diff
@@ -444,8 +518,8 @@ private function rowToPrincipal(array $row, array $metadata = []):array {
 	}
 
 	/**
-	 * @param $row
-	 * @param $userGroups
+	 * @param array $row
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:521]:
```suggestion
	 * @param string[] $row
```

File: apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php [hunk 8]
```diff
@@ -444,8 +518,8 @@ private function rowToPrincipal(array $row, array $metadata = []):array {
...
+	 * @param array $userGroups
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/ResourceBooking/AbstractPrincipalBackend.php:522]:
```suggestion
	 * @param string[] $userGroups
```

**Review by ChristophWurst — DISMISSED** [review]
Let's narrow down the types. Rest looks :fire:

**Comment by miaulalala** [comment]
StatAna isn't liking all of them :( will revert

**Review by ChristophWurst — APPROVED** [review]

**Comment by ChristophWurst** [comment]
Drone failure is unrelated

**Pull request merged**